chips/sam4l/src/usart.rs

```diff
@@ -322,30 +533,31 @@ impl USART {
         // stopped it from occurring just in case it caused issues in the
         // future.
         if self.is_clock_enabled() {
-            let regs: &USARTRegisters = unsafe { &*self.registers };
+            let regs: &UsartRegisters = unsafe { &*self.registers };
             let status = regs.csr.get();
             let mask = regs.imr.get();
 
             // Reset status registers. We need to do this first because some
             // interrupts signal us to turn off our clock.
-            regs.cr.set(1 << 8); // RSTSTA
+            regs.cr.write(Control::RSTSTA.val(1));
 
-            if status & (1 << 8) != 0 && mask & (1 << 8) != 0 {
+            if (status & ChannelStatus::TIMEOUT.val(1).to_u32()) != 0
```

Comment:
I really don't like escaping the register mechanism and doing integer comparison. I know this is a particularly antagonistic piece of code, but I think we can commit the lesser sin of code duplication instead. What if we reset the status registers inside each branch of the if statement?

The code would look like (comments elided):
```
let regs: &UsartRegisters = unsafe { &*self.registers };

if regs.csr.matches(ChannelStatus::Timeout::SET) && regs.imr.matches(Interrupt::TIMEOUT::SET) {
    regs.cr.write(Control::RSTSTA::SET);
    self.disable_rx_timeout();
    self.abort_rx(hil::uart::Error::CommandComplete);
} else if regs.csr.matches(ChannelStatus::TXEMPTY::SET) && regs.imr.matches(Interrupt::TXEMPTY::SET) {
    regs.cr.write(Control::RSTSTA::SET);
    self.disable_tx_empty_interrupt();
    self.disable_tx();
    self.usart_tx_state.set(USARTStateTX::Idle);
} else ...
```

Reply:
I'll try.